### PR TITLE
Aha/analytics

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/analytics/events/Event.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * <p>
+ * Sub-classes will be persisted in the <tt>analytics</tt> database realm which is expected to
+ * point to a <b>Clickhouse</b> database.
+ */
+@Realm("analytics")
+@Engine("MergeTree() PARTITION BY toYYYYMM(eventDate) ORDER BY (eventDate, eventTimestamp)")
+public abstract class Event extends SQLEntity {
+
+    /**
+     * Represents an internal in-memory flag which determines if storing the entry has already been retried.
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/analytics/events/Event.java]: -storing

Comment (Tobsucht) [reply src/main/java/sirius/biz/analytics/events/Event.java]: typo retrieved

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Responsible for collecting and storing {@link Event events} for analytical and statistical purposes.
+ * <p>
+ * To minimize the impact on the running application and to maximize the performance, events are collected and queue.
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]: queued

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            buffer.offer(event);
+        } catch (HandledException e) {
+            Exceptions.ignore(e);
+        } catch (Exception e) {
+            Exceptions.handle(Log.BACKGROUND, e);
+        }
+    }
+
+    /**
+     * Invoked periodically by the {@link EventProcessorLoop} to process events if necessarry.
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]: typo necessary

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+     * <p>
+     * An insertion run will be started if there are enough events in the buffer (more than {@link #MIN_BUFFER_SIZE})
+     * or if enough time elapsed since the last insertion run (more than {@link #MAX_BUFFER_AGE}).
+     *
+     * @return <tt>true</tt> if events where processed, <tt>false</tt> otherwise
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]: return is int, not boolean

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+        return processedEvents;
+    }
+
+    /**
+     * Inserts a single event by either creating a new {@link InsertQuery} or by appending a batch-insert for an
+     * existing once (there is one per event type in <tt>queries</tt>).
+     *
+     * @param ctx     the batch context
+     * @param queries the map of perpared insert queries per event type
```
Comment (Tobsucht) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]: typo prepared

File: src/main/java/sirius/biz/analytics/events/Event.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    /**
+     * Contains the date when the event occurred.
+     * <p>
+     * Note that this is redundant to {@link #eventTimestamp} and only required to be used as index by <b>Clickhouse</b>.
+     */
+    private LocalDate eventDate;
+
+    /**
+     * Contains the date and them when the event occurred.
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/events/Event.java]:
```suggestion
     * Contains the date and time when the event occurred.
```

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java (hunk 4 above)
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]:
```suggestion
     * @return the number of inserted events
```

File: src/main/java/sirius/biz/analytics/events/EventRecorder.java [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        } catch (Exception e) {
+            Exceptions.handle(Log.BACKGROUND, e);
+        }
+
+        return processedEvents;
+    }
+
+    /**
+     * Inserts a single event by either creating a new {@link InsertQuery} or by appending a batch-insert for an
+     * existing once (there is one per event type in <tt>queries</tt>).
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/events/EventRecorder.java]:
```suggestion
     * existing one (there is one per event type in <tt>queries</tt>).
```

**Pull request merged**